### DISCUSSION TITLE
pnpm audit returns "Cannot read property 'metadata' of undefined"

### DISCUSSION THREAD

**Discussion by totojack** [pnpm/pnpm#3920] [Q&A]
Hello.

We're trying to audit the project dependencies. But the command returns this error message:

> ERROR  Cannot read property 'metadata' of undefined

The pnpm version used is 6.19.0.
I tried the same using a previous version (6.18.0) and there is a stack trace after the error message:

> ERROR  Cannot read property 'metadata' of undefined
> at handler [as audit]         ../usr/local/lib/node_modules/pnpm/dist/pnpm.cjs:88949       const vulnerabilities = …
> at processTicksAndRejections  node:internal/process/task_queues:96
> at                            async /usr/local/lib/node_modules/pnpm/dist/pnpm.cjs:133538
> at async run                  ../usr/local/lib/node_modules/pnpm/dist/pnpm.cjs:133513      let { output, exitCode }…
> at async runPnpm              ../usr/local/lib/node_modules/pnpm/dist/pnpm.cjs:133724      await main(argv);
> at                            async /usr/local/lib/node_modules/pnpm/dist/pnpm.cjs:133716

The line

> /usr/local/lib/node_modules/pnpm/dist/pnpm.cjs:88949

has this content

> const vulnerabilities = auditReport.metadata.vulnerabilities;

Could it be a problem with our npm registry not returning an audit report?

Thanks.

**Comment by InsOpDe** [comment]
Had the same problem and tried: 
```
 pnpm audit --audit-level high --ignore-registry-errors
The audit endpoint (at http://registry.npmjs.org/-/npm/v1/security/audits) responded with 426: Beginning October 4, 2021, all connections to npm websites and the npm registry - including for package installation - must use HTTPS and TLS 1.2 or higher.
```
then I looked at my `.npmrc` and saw, that I entered a registry without `https`:
```
registry=http://registry.npmjs.org/
```

fixed it with:
```
registry=https://registry.npmjs.org/
```

now it works again

**Comment by totojack** [comment]
I tried it after a pnpm update. The error message is now a bit different:

>ERROR  Cannot read properties of undefined (reading 'metadata')

Enabling the --ignore-registry-errors parameter I have a clear detail:

```
The audit endpoint (at https://xxx) responded with 401: {
  "errors" : [ {
    "status" : 401,
    "message" : "Unauthorized"
  } ]
}
```

So I guess I must check with user permissions on the endpoint to be able to audit packages. It's a custom Artifactory service.

**Reply by sled** [reply]
Did you have any luck? If I use the plain old `npm audit` it seems to work, however with pnpm I get a 401 too on my private registry.

**Reply by totojack** [reply]
Unfortunately I'm not working anymore on this specific project.

**Reply by sled** [reply]
Thanks for the response 👍 I tracked down the issue, pnpm-audit didn't use the credentials, opened up a PR #5053

**Comment by sled** [comment]
I'm having the same issue, see https://github.com/pnpm/pnpm/issues/5038
